File: app/src/main/java/com/scatl/uestcbbs/module/board/view/BoardPostFragment.java
package com.scatl.uestcbbs.module.board.view;


import android.content.Intent;
import android.content.res.ColorStateList;
import android.graphics.Color;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import android.view.ContextThemeWrapper;
import android.view.LayoutInflater;
import android.view.View;
import android.view.animation.AnimationUtils;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.google.android.material.chip.Chip;
import com.google.android.material.chip.ChipGroup;
import com.scatl.uestcbbs.R;
import com.scatl.uestcbbs.annotation.ToastType;
import com.scatl.uestcbbs.api.ApiConstant;
import com.scatl.uestcbbs.base.BaseFragment;
import com.scatl.uestcbbs.base.BasePresenter;
import com.scatl.uestcbbs.callback.HapticClickListener;
import com.scatl.uestcbbs.callback.OnRefresh;
import com.scatl.uestcbbs.entity.CommonPostBean;
import com.scatl.uestcbbs.module.post.adapter.CommonPostAdapter;
import com.scatl.uestcbbs.module.webview.view.WebViewActivity;
import com.scatl.uestcbbs.widget.MyLinearLayoutManger;

import com.scatl.uestcbbs.module.board.presenter.BoardPostPresenter;
import com.scatl.uestcbbs.module.post.view.NewPostDetailActivity;
import com.scatl.uestcbbs.module.user.view.UserDetailActivity;
import com.scatl.uestcbbs.util.Constant;
import com.scatl.uestcbbs.util.RefreshUtil;
import com.scatl.uestcbbs.util.SharePrefUtil;
import com.scatl.util.ScreenUtil;
import com.scwang.smart.refresh.layout.SmartRefreshLayout;
import com.scwang.smart.refresh.layout.api.RefreshLayout;
import com.scwang.smart.refresh.layout.constant.RefreshState;

public class BoardPostFragment extends BaseFragment implements BoardPostView{

    private SmartRefreshLayout refreshLayout;
    private RecyclerView recyclerView;
    private CommonPostAdapter boardPostAdapter;
    private TextView hint, errorText;
    private LinearLayout error500Layout;
    private Button openBrowserBtn;
    private View filterView;
    private ChipGroup mChipGroup;

    private int boardId, mFid, page = 1;
    private String sortBy;

    private BoardPostPresenter boardPostPresenter;

    public static BoardPostFragment getInstance(Bundle bundle) {
        BoardPostFragment boardPostFragment = new BoardPostFragment();
        boardPostFragment.setArguments(bundle);
        return boardPostFragment;
    }

    @Override
    protected void getBundle(Bundle bundle) {
        super.getBundle(bundle);
        if (bundle != null) {
            boardId = bundle.getInt(Constant.IntentKey.BOARD_ID, Integer.MAX_VALUE);
            mFid = bundle.getInt(Constant.IntentKey.FILTER_ID, Integer.MAX_VALUE);
            sortBy = bundle.getString(Constant.IntentKey.TYPE);
        }
    }

    @Override
    protected int setLayoutResourceId() {
        return R.layout.fragment_board_post;
    }

    @Override
    protected void findView() {
        recyclerView = view.findViewById(R.id.fragment_board_post_rv);
        refreshLayout = view.findViewById(R.id.fragment_board_post_refresh);
        hint = view.findViewById(R.id.fragment_board_post_hint);
        error500Layout = view.findViewById(R.id.fragment_board_error_500_layout);
        openBrowserBtn = view.findViewById(R.id.fragment_board_open_browser_btn);
        errorText = view.findViewById(R.id.fragment_board_error_text);

        filterView = LayoutInflater.from(mActivity).inflate(R.layout.view_board_post_filter, null, false);
        mChipGroup = filterView.findViewById(R.id.chip_group);
    }

    @Override
    protected void initView() {
        boardPostPresenter = (BoardPostPresenter) presenter;

        openBrowserBtn.setOnClickListener(this);

        boardPostAdapter = new CommonPostAdapter(R.layout.item_common_post, "", null);
        boardPostAdapter.addHeaderView(filterView, 0);
        recyclerView.setLayoutManager(new MyLinearLayoutManger(mActivity));
        recyclerView.setAdapter(boardPostAdapter);
        recyclerView.setLayoutAnimation(AnimationUtils.loadLayoutAnimation(mActivity, R.anim.layout_animation_scale_in));

    }

    @Override
    protected void lazyLoad() {
        super.lazyLoad();
        refreshLayout.autoRefresh(10, 300, 1, false);
    }

    @Override
    protected BasePresenter initPresenter() {
        return new BoardPostPresenter();
    }

    @Override
    protected void onClickListener(View v) {
        if (v.getId() == R.id.fragment_board_open_browser_btn) {
            if (openBrowserBtn.getTag().equals(ErrorStatus.STATUS_500)) {
                Intent intent = new Intent(mActivity, WebViewActivity.class);
                intent.putExtra(Constant.IntentKey.URL, ApiConstant.Post.BOARD_URL + boardId);
                startActivity(intent);
            } else if (openBrowserBtn.getTag().equals(ErrorStatus.STATUS_NEED_PAY)) {
                boardPostPresenter.payForVisiting(boardId, SharePrefUtil.getForumHash(mActivity));
            }
        }
    }

    @Override
    protected void setOnItemClickListener() {
        boardPostAdapter.setOnItemChildClickListener((adapter, view, position) -> {
            if (view.getId() == R.id.avatar) {
                Intent intent = new Intent(mActivity, UserDetailActivity.class);
                intent.putExtra(Constant.IntentKey.USER_ID, boardPostAdapter.getData().get(position).user_id);
                startActivity(intent);
            }
            if (view.getId() == R.id.content_layout) {
                Intent intent = new Intent(mActivity, NewPostDetailActivity.class);
                intent.putExtra(Constant.IntentKey.TOPIC_ID, boardPostAdapter.getData().get(position).topic_id);
                startActivity(intent);
            }
        });
    }

    @Override
    protected void setOnRefreshListener() {
        RefreshUtil.setOnRefreshListener(mActivity, refreshLayout, new OnRefresh() {
            @Override
            public void onRefresh(RefreshLayout refreshLayout) {
                page = 1;
                boardPostPresenter.getBoardPostList(page,
                        SharePrefUtil.getPageSize(mActivity), 1,
                        boardId, mFid, "typeid", sortBy,
                        mActivity);
            }

            @Override
            public void onLoadMore(RefreshLayout refreshLayout) {
                boardPostPresenter.getBoardPostList(page,
                        SharePrefUtil.getPageSize(mActivity), 1,
                        boardId, mFid, "typeid", sortBy,
                        mActivity);
            }
        });
    }

    @Override
    public void onGetBoardPostSuccess(CommonPostBean singleBoardBean) {
        page = page + 1;

        error500Layout.setVisibility(View.GONE);

        if (refreshLayout.getState() == RefreshState.Refreshing) {
            if (singleBoardBean.has_next == 1) {
                refreshLayout.finishRefresh();
            } else {
                refreshLayout.finishRefreshWithNoMoreData();
            }
        }
        if (refreshLayout.getState() == RefreshState.Loading) {
            if (singleBoardBean.has_next == 1) {
                refreshLayout.finishLoadMore(true);
            } else {
                refreshLayout.finishLoadMoreWithNoMoreData();
            }
        }

        if (singleBoardBean.page == 1) {
            boardPostAdapter.addData(singleBoardBean.list, true);
            setFilterView(singleBoardBean);
            recyclerView.scheduleLayoutAnimation();
        } else {
            boardPostAdapter.addData(singleBoardBean.list, false);
        }

        hint.setText(boardPostAdapter.getData().size() == 0 ? "啊哦，这里空空的" : "");
    }

    private void setFilterView(CommonPostBean singleBoardBean) {
        if (singleBoardBean.classificationType_list != null && singleBoardBean.classificationType_list.size() > 0) {
            mChipGroup.removeAllViews();
            mChipGroup.addView(getChip("全部", 0));
            for (int i = 0; i < singleBoardBean.classificationType_list.size(); i ++) {
                Chip chip = getChip(singleBoardBean.classificationType_list.get(i).classificationType_name,
                        singleBoardBean.classificationType_list.get(i).classificationType_id);
                mChipGroup.addView(chip);
            }
            filterView.setVisibility(View.VISIBLE);
        } else {
            filterView.setVisibility(View.GONE);
        }
    }

    private Chip getChip(String text, final int filterId) {
        Chip chip = new Chip(new ContextThemeWrapper(mActivity, R.style.Widget_Material3_Chip_Filter));
        chip.setText(text);
        chip.setCheckable(true);
        chip.setChipStrokeWidth(0);
        chip.setChipCornerRadius(ScreenUtil.dip2px(mActivity, 25));
        chip.setChipStrokeColor(ColorStateList.valueOf(Color.parseColor("#00000000")));
        chip.setChecked(filterId == mFid);
        chip.setOnClickListener(new HapticClickListener() {
            @Override
            public void onViewClick(@NonNull View v) {
                Chip c = (Chip) v;
                if (c.isChecked()) {
                    mFid = filterId;
                    recyclerView.scrollToPosition(0);
                    refreshLayout.autoRefresh(10, 300, 1, false);
                } else {
                    c.setChecked(true);
                }
            }
        });
        return chip;
    }

    @Override
    public void onGetBoardPostError(String msg) {
        if (refreshLayout.getState() == RefreshState.Refreshing) {
            refreshLayout.finishRefresh();
        }
        if (refreshLayout.getState() == RefreshState.Loading) {
            refreshLayout.finishLoadMore(false);
        }

        if (msg.contains(ApiConstant.Code.RESPONSE_ERROR_500)) {
            errorText.setText(getString(R.string.board_error_500));
            openBrowserBtn.setText("打开Web页面");
            openBrowserBtn.setTag(ErrorStatus.STATUS_500);
            error500Layout.setVisibility(View.VISIBLE);
        } else if (msg.contains("您需要支付")){
            errorText.setText(msg);
            openBrowserBtn.setTag(ErrorStatus.STATUS_NEED_PAY);
            openBrowserBtn.setText("立即支付");
            error500Layout.setVisibility(View.VISIBLE);
        } else {
            error500Layout.setVisibility(View.GONE);
            hint.setText(msg);
        }
    }

    @Override
    public void onPaySuccess(String msg) {
        error500Layout.setVisibility(View.GONE);
        showToast(msg, ToastType.TYPE_SUCCESS);
    }

    @Override
    public void onPayError(String msg) {
        showToast(msg, ToastType.TYPE_ERROR);
    }

    private enum ErrorStatus {
        STATUS_NEED_PAY,
        STATUS_500
    }
}

File: app/src/main/java/com/scatl/uestcbbs/module/board/view/BoardPostFragment.kt
package com.scatl.uestcbbs.module.board.view

import android.content.Intent
import android.content.res.ColorStateList
import android.graphics.Color
import android.os.Bundle
import android.view.ContextThemeWrapper
import android.view.HapticFeedbackConstants
import android.view.View
import android.view.animation.AnimationUtils
import com.google.android.material.chip.Chip
import com.scatl.uestcbbs.R
import com.scatl.uestcbbs.annotation.PostSortByType
import com.scatl.uestcbbs.annotation.ToastType
import com.scatl.uestcbbs.base.BaseVBFragment
import com.scatl.uestcbbs.databinding.FragmentBoardPostBinding
import com.scatl.uestcbbs.entity.CommonPostBean
import com.scatl.uestcbbs.module.board.presenter.BoardPostPresenter
import com.scatl.uestcbbs.module.post.adapter.CommonPostAdapter
import com.scatl.uestcbbs.module.post.view.NewPostDetailActivity
import com.scatl.uestcbbs.module.user.view.UserDetailActivity
import com.scatl.uestcbbs.util.Constant
import com.scatl.uestcbbs.util.SharePrefUtil
import com.scatl.uestcbbs.util.showToast
import com.scatl.util.ScreenUtil
import com.scwang.smart.refresh.layout.api.RefreshLayout

/**
 * created by sca_tl at 2023/6/10 12:06
 */
class BoardPostFragment: BaseVBFragment<BoardPostPresenter, BoardPostView, FragmentBoardPostBinding>(), BoardPostView {

    private lateinit var mCommonPostAdapter: CommonPostAdapter

    private var mBoardId = 0
    private var mFid = 0
    private var mPage = 1
    private var mSortBy = PostSortByType.TYPE_ALL
    private var mNoMoreData = false

    companion object {
        fun getInstance(bundle: Bundle?) = BoardPostFragment().apply { arguments = bundle }
    }

    override fun getBundle(bundle: Bundle?) {
        bundle?.let {
            mBoardId = it.getInt(Constant.IntentKey.BOARD_ID, Int.MAX_VALUE)
            mFid = it.getInt(Constant.IntentKey.FILTER_ID, Int.MAX_VALUE)
            mSortBy = it.getString(Constant.IntentKey.TYPE, PostSortByType.TYPE_ALL)
        }
    }

    override fun getViewBinding() = FragmentBoardPostBinding.inflate(layoutInflater)

    override fun initPresenter() = BoardPostPresenter()

    override fun initView() {
        super.initView()

        bindClickEvent(mBinding.payBtn)

        mCommonPostAdapter = CommonPostAdapter("", onPreload = {
            if (SharePrefUtil.isAutoLoadMore(context) && !mNoMoreData) {
                lazyLoad()
            }
        })
        mBinding.recyclerView.adapter = mCommonPostAdapter
        mBinding.recyclerView.layoutAnimation = AnimationUtils.loadLayoutAnimation(context, R.anim.layout_animation_scale_in)

        mBinding.statusView.loading(mBinding.filterLayout)
    }

    override fun lazyLoad() {
        mPresenter?.getBoardPostList(
            mPage, SharePrefUtil.getPageSize(context), 1,
            mBoardId, mFid, "typeid", mSortBy
        )
    }

    override fun onClick(v: View) {
        super.onClick(v)
        when(v) {
            mBinding.payBtn -> {
                mPresenter?.payForVisiting(mBoardId, SharePrefUtil.getForumHash(context))
            }
        }
    }

    override fun setOnItemClickListener() {
        mCommonPostAdapter.addOnItemChildClickListener(R.id.avatar) { adapter, view, position ->
            val intent = Intent(context, UserDetailActivity::class.java).apply {
                putExtra(Constant.IntentKey.USER_ID, mCommonPostAdapter.items[position].user_id)
            }
            startActivity(intent)
        }
        mCommonPostAdapter.addOnItemChildClickListener(R.id.content_layout) { adapter, view, position ->
            val intent = Intent(context, NewPostDetailActivity::class.java).apply {
                putExtra(Constant.IntentKey.TOPIC_ID, mCommonPostAdapter.items[position].topic_id)
            }
            startActivity(intent)
        }
    }

    override fun onRefresh(refreshLayout: RefreshLayout) {
        mPage = 1
        lazyLoad()
    }

    override fun onLoadMore(refreshLayout: RefreshLayout) {
        if (!mCommonPostAdapter.isPreloading) {
            lazyLoad()
        }
    }

    override fun onGetBoardPostSuccess(commonPostBean: CommonPostBean) {
        mCommonPostAdapter.isPreloading = false
        mBinding.statusView.success()
        mBinding.refreshLayout.finishRefresh()

        if (mPage == 1) {
            setFilterView(commonPostBean)
            if (commonPostBean.list.isEmpty()) {
                mBinding.statusView.error("啊哦，这里空空的~")
            } else {
                mCommonPostAdapter.addData(commonPostBean.list, true)
                mBinding.recyclerView.scheduleLayoutAnimation()
            }
        } else {
            mCommonPostAdapter.addData(commonPostBean.list, false)
        }

        if (commonPostBean.has_next == 1) {
            mPage ++
            mBinding.refreshLayout.finishLoadMore(true)
        } else {
            mNoMoreData = true
            mBinding.refreshLayout.finishLoadMoreWithNoMoreData()
        }
    }

    override fun onGetBoardPostError(msg: String?) {
        mCommonPostAdapter.isPreloading = false
        mBinding.refreshLayout.finishRefresh()
        if (mPage == 1) {
            if (mCommonPostAdapter.items.isNotEmpty()) {
                showToast(msg, ToastType.TYPE_ERROR)
            } else {
                mBinding.statusView.error(msg)
                if (msg?.contains("您需要支付") == true) {
                    mBinding.payBtn.visibility = View.VISIBLE
                }
            }
            mBinding.refreshLayout.finishLoadMore()
        } else {
            mBinding.refreshLayout.finishLoadMore(false)
        }
    }

    override fun onPaySuccess(msg: String?) {
        mBinding.payBtn.visibility = View.GONE
        lazyLoad()
        showToast(msg, ToastType.TYPE_SUCCESS)
    }

    override fun onPayError(msg: String?) {
        showToast(msg, ToastType.TYPE_ERROR)
    }

    private fun setFilterView(singleBoardBean: CommonPostBean) {
        if (singleBoardBean.classificationType_list != null && singleBoardBean.classificationType_list.size > 0) {
            mBinding.chipGroup.removeAllViews()
            mBinding.chipGroup.addView(getChip("全部", 0))
            for (i in singleBoardBean.classificationType_list.indices) {
                val chip = getChip(
                    singleBoardBean.classificationType_list[i].classificationType_name,
                    singleBoardBean.classificationType_list[i].classificationType_id
                )
                mBinding.chipGroup.addView(chip)
            }
            mBinding.filterLayout.visibility = View.VISIBLE
        } else {
            mBinding.filterLayout.visibility = View.GONE
        }
    }

    private fun getChip(text: String, filterId: Int): Chip {
        return Chip(ContextThemeWrapper(context, R.style.Widget_Material3_Chip_Filter)).apply {
            this.text = text
            isCheckable = true
            chipStrokeWidth = 0f
            chipCornerRadius = ScreenUtil.dip2px(requireContext(), 25f).toFloat()
            chipStrokeColor = ColorStateList.valueOf(Color.parseColor("#00000000"))
            isChecked = filterId == mFid
            setOnClickListener {
                it.performHapticFeedback(HapticFeedbackConstants.CONTEXT_CLICK)
                val c = it as Chip
                if (c.isChecked) {
                    mFid = filterId
                    mBinding.recyclerView.scrollToPosition(0)
                    mBinding.refreshLayout.autoRefresh(10, 300, 1f, false)
                } else {
                    c.isChecked = true
                }
            }
        }
    }
}